Silicon/Broadcom/Net: Add Genet stub driver to setup MAC

pbatard opened the pull request:
The RPi4 has a Gigabit Broadcom Genet platform device. In order to
assist in booting the platform in an ACPI environment, it is
desirable if the hardware can describe itself rather than passing
parameters via DSDT/DSD. One way to achieve this is to assure that
the adapter is preprogrammed with the correct ethernet MAC address
which, on the Raspberry Pi, can be achieved by using a dedicated
firmware interface call.

The driver is added under Silicon on account that Genet NICs are
not specific to the Raspberry Pi and could be extended for other
platforms. This is also why the dedicated PcdBcmGenetRpiMacInit
PCD is introduced for the RPi specific MAC initialization.

Signed-off-by: Pete Batard <pete@akeo.ie>

jlinton committed:
Silicon/Broadcom/Net: Add Genet stub driver to setup MAC

Add a stub for the Broadcom Genet network interface, that is used
by the Raspberry Pi 4 (and that may be used by other platforms).

For the time being, the stub only performs UMAC init, using a MAC
address PCD, as, even without a working network interface in UEFI
environment, it is desirable if the hardware can describe itself
for booting in an ACPI environment, rather than pass parameters via
DSDT/DSD .

Signed-off-by: Pete Batard <pete@akeo.ie>

pbatard commented:
This is Jeremy's #7 reworked to reside in Silicon. Note that I have tested compilation & boot, but not actual usage from kernel, so it's possible that it still needs some tweaking.

Besides `PcdBcmGenetRegistersAddress` I'm also introducing `PcdBcmGenetRpiMacInit` since the init we perform through RPiFw protocol is of course specific to that platform and, with the driver being in Silicon, we need to allow other platforms to perform alternate MAC init.

pbatard commented on Silicon/Broadcom/Drivers/Net/BcmNet.dec:
```diff
@@ -0,0 +1,23 @@
+  PACKAGE_VERSION                = 1.0
+
+#[Includes]
+#  Include
```
Forgot to remove this commented section. I'll do so when upstreaming.

ardbiesheuvel commented:
On Thu, 16 Jan 2020 at 18:19, Pete Batard <notifications@github.com> wrote:
>
> @pbatard commented on this pull request.
>
> ________________________________
>
> In Silicon/Broadcom/Drivers/Net/BcmNet.dec:
>
> > +## @file
> +#
> +#  Copyright (c) 2020, Pete Batard <pete@akeo.ie>
> +#
> +#  SPDX-License-Identifier: BSD-2-Clause-Patent
> +#
> +##
> +
> +[Defines]
> +  DEC_SPECIFICATION              = 0x0001001A
> +  PACKAGE_NAME                   = BcmNetPkg
> +  PACKAGE_GUID                   = 34E19823-D23A-41AB-9C09-ED1225B32DFF
> +  PACKAGE_VERSION                = 1.0
> +
> +#[Includes]
> +#  Include
>
> Forgot to remove this commented section. I'll do so when upstreaming.
>
>

Apologies for not bringing this up immediately when you suggested it,
but I fail to see the point of making this a 'generic' driver in
Silicon/Broadcom while it is so closely tied to the RPi firmware
protocol, and doesn't do anything useful without it.

So this should simply be a RPi platform DXE driver, either a
standalone one for genet or one that will assume other RPI4 specific
duties going forward.

pbatard commented:
Okay. My understanding is that this is a genet stub we will build upon, i.e. right now it's only used for UMAC init but it's going to become a full fledged genet driver (with other platforms potentially doing their own UMAC init). So that's how I treated it, since I'd rather not upstream something into `Platform`, only to be told that it should be better moved to `Silicon` a while later...

I guess the alternative is that this should to be called `Genet` at all, but become a dedicated `UmacDxe` or something, to indicate that this is not going to be built upon for genet..

Also, nothwithstanding this, because we're going to have genet anyway, which I don't think can reside anywhere but in `Silicon`, we're still going to want some kind of genet header there at some stage, which we can't have under `Bcm27xx` or `Bcm283x` because it's not specific to those. But I guess we can leave that for until we see a bit clearer, and just have that `Genet.h` in `Platform/.../UmacDxe` for the time being...

ardbiesheuvel commented:
On Fri, 17 Jan 2020 at 11:00, Pete Batard <notifications@github.com> wrote:
>
> Okay. My understanding is that this is a genet stub we will build upon, i.e. right now it's only used for UMAC init but it's going to become a full fledged genet driver (with other platforms potentially doing their own UMAC init). So that's how I treated it, since I'd rather not upstream something into Platform, only to be told that it should be better moved to Silicon a while later...
>
> I guess the alternative is that this should to be called Genet at all, but become a dedicated UmacDxe or something, to indicate that this is not going to be built upon for genet..
>
> Also, nothwithstanding this, because we're going to have genet anyway, which I don't think can reside anywhere but in Silicon, we're still going to want some kind of genet header there at some stage, which we can't have under Bcm27xx or Bcm283x because it's not specific to those. But I guess we can leave that for until we see a bit clearer, and just have that Genet.h in Platform/.../UmacDxe for the time being...
>

If you want to stick with approach, you should at least factor out the
MAC discovery using a library class, and provide an implementation of
that library class for the RPi4 platform that invokes the RPi firmware
protocol, and not rely on #ifdef's etc and have a code path that does
nothing.

pbatard commented:
Yes, I agree. I don't like having those #ifdefs either but it seemed a bit like an overkill to go that far...

I'll see if I want to take another stab at this. If not, I'll just go for `UmacDxe`.

pbatard committed:
Platform/RPi: Add PlatformPcdLib to set the Genet MAC address

The Genet driver stub used by the Raspberry Pi 4 platform is
designed to set the MAC address according to a PCD.

To be able to set that PCD at runtime, by using the Raspberry
Pi firmware interface, that has a dedicated call to retrieve
the MAC address, and satisfy driver dependencies in a generic
manner, we create a new PlatformPcdLib that can be referenced
by the Genet driver, to set the MAC PCD before use there.

While it is currently only tailored around MAC PCD population
for Genet, we do expect this PCD library to be extended in the
future, to provide additional PCD facilities for other drivers.

Signed-off-by: Pete Batard <pete@akeo.ie>

jlinton committed:
Platform/Rpi4: Enable Broadcom Genet stub driver

This adds the required references to use the Genet stub driver in
order to dynamically populate the MAC address for OS consumption.

Signed-off-by: Pete Batard <pete@akeo.ie>

pbatard commented:
Updated version of what I am planning to upstream, unless someone disagrees.

This series now uses a PCD to hold the MAC address (which has the advantage that someone who wants to force a MAC during build can do so) that gets updated at runtime, through a new `PlatformPcdLib` library that fetches the value from the `RpiFirmware` interface.

After looking at various approaches, I think this is the cleanest & most efficient way of achieving UMAC init, whilst enabling us to have a Genet stub we can build on.

jlinton commented:
This all looks fine to me. Since I didn't see any code to allow the user to override the mac address, for a future todo, it might be nice if the user could set it via the BDS menus.

pbatard commented:
> it might be nice if the user could set it via the BSD menus

Yeah, I've been thinking about doing that too, as well as, if we see any use for this, have an option that users could check to set a random MAC on each boot.

I've added a note for user MAC specification in conf to the task list.

jlinton commented on Platform/RaspberryPi/Library/PlatformPcdLib/PlatformPcdLib.c:
```diff
@@ -0,0 +1,59 @@
+  //
+  Status = mFwProtocol->GetMacAddress ((UINT8*) &MacAddr);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_WARN, "%a: failed to retrieve MAC address\n", __FUNCTION__));
```
Seems like this should be doing something with the value of PcdBcmGenetMacAddress in the face of failure.

pbatard replied:
Hmm, you're right that there's an issue here.

There's an `if (PcdGet64 (PcdBcmGenetMacAddress) == 0)` that's hinted at in the comment, but that I forgot to add in this version of the code, and that should have guarded the whole MAC address retrieval and PCD updating so that we don't overwrite a PCD that's already been set by the user. I'll fix that in the code I upstream (probably tomorrow).

However, I don't see why or how we could do something with `PcdBcmGenetMacAddress` if `GetMacAddress ()` fails once we have the missing `if (PcdGet64 (PcdBcmGenetMacAddress) == 0)` guard, because then the PCD can only be zero, and we won't have a MAC to work with, so the best we could do here would be to set a random value. However, considering that the Linux drivers will detect if the UMAC has not been set and assign a random MAC value then, and I would suggest we do the same in our UEFI Genet driver (since I don't think we want to assume that UMAC has always been properly set), I'd assert that the best course of action is to just do nothing.

Or do you actually rely on the stub UMAC initialization to be performed always, even if the MAC is zero? Coz the way I saw it is that this part of the Genet code is only meant to be used for MAC setup if we have one, and that the rest of the driver would know how to deal with UMAC regardless of whether `SetMacAddress ()` has been called or not.

I guess this is something we can sort out once we start adding the actual network driver code, but it may help to clarify what you had in mind with UMAC init.

jlinton replied:
The current code, (i'm about to post) does set a random mac if the one read from the adapter is invalid (all 0, 0xff or a multicast). But it must have the umac taken out of reset. So I guess its ok for the time being to be 0 as long as the adapter isn't in reset.

But as you mention, that won't fly in for a future fully featured driver.

pbatard commented:
Closing this PR as it is being upstreamed as [53437](https://edk2.groups.io/g/devel/message/53437), [53438](https://edk2.groups.io/g/devel/message/53438), [53439](https://edk2.groups.io/g/devel/message/53439) and [53442](https://edk2.groups.io/g/devel/message/53442).
